src/filter/bgsubtract0r/bgsubtract0r.c

```diff
@@ -0,0 +1,254 @@
+} bgsubtract0r_instance_t;
+
+int f0r_init()
+{
+  return 1;
+}
+
+void f0r_deinit()
+{
+}
+
+void f0r_get_plugin_info(f0r_plugin_info_t* bgsubtract0r_info)
+{
+  bgsubtract0r_info->name = "BgSubtract0r";
+  bgsubtract0r_info->author = "Samuel Mimram";
+  bgsubtract0r_info->plugin_type = F0R_PLUGIN_TYPE_FILTER;
+  bgsubtract0r_info->color_model = F0R_COLOR_MODEL_RGBA8888;
+  bgsubtract0r_info->frei0r_version = FREI0R_MAJOR_VERSION;
+  bgsubtract0r_info->major_version = 0;
+  bgsubtract0r_info->minor_version = 2;
+  bgsubtract0r_info->num_params =  3;
+  bgsubtract0r_info->explanation = "Bluescreen the background of a static video.";
+}
+
+f0r_instance_t f0r_construct(unsigned int width, unsigned int height)
+{
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)calloc(1, sizeof(*inst));
+  inst->width = width;
+  inst->height = height;
+  inst->denoise = 1;
+  inst->blur = 0;
+  inst->threshold = 26;
+  inst->reference = NULL;
+  inst->mask = malloc(width*height);
+  return (f0r_instance_t)inst;
+}
+
+void f0r_destruct(f0r_instance_t instance)
+{
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  if (inst->reference) free(inst->reference);
+  free(inst->mask);
+  free(inst);
+}
+
+void f0r_get_param_info(f0r_param_info_t* info, int param_index)
+{
+  switch(param_index)
+  {
+  case 0:
+    info->name = "threshold";
+    info->type = F0R_PARAM_DOUBLE;
+    info->explanation = "Threshold for difference";
+    break;
+
+  case 1:
+    info->name = "denoise";
+    info->type = F0R_PARAM_BOOL;
+    info->explanation = "Remove noise";
+    break;
+
+  case 2:
+    info->name = "blur";
+    info->type = F0R_PARAM_DOUBLE;
+    info->explanation = "Blur alpha channel by given radius (to remove sharp edges)";
+    break;
+  }
+}
+
+void f0r_set_param_value(f0r_instance_t instance, f0r_param_t param, int param_index)
+{
+  assert(instance);
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  f0r_param_color_t *c;
+
+  switch(param_index)
+  {
+  case 0:
+    inst->threshold = *((double*)param) * 255.;
+    break;
+
+  case 1:
+    inst->denoise = *((double*)param) >= 0.5;
+    break;
+
+  case 2:
+    inst->blur = (int)(*((double*)param)+0.5);
+    break;
+  }
+}
+
+void f0r_get_param_value(f0r_instance_t instance, f0r_param_t param, int param_index)
+{
+  assert(instance);
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  f0r_param_color_t *c;
+
+  switch(param_index)
+  {
+  case 0:
+    *((double*)param) = (double)inst->threshold / 255.;
+    break;
+
+  case 1:
+    *((double*)param) = inst->denoise ? 1. : 0.;
+    break;
+
+  case 2:
+    *((double*)param) = inst->blur;
+    break;
+  }
+}
+
+#define MAX(x,y) ((x) > (y) ? (x) : (y));
+
+inline static uint8_t dst(uint32_t x, uint32_t y)
+{
+  uint8_t d;
+
+  d = abs((x&0xff)-(y&0xff));
+  d = MAX(d, abs(((x>>8)&0xff)-((y>>8)&0xff)));
+  d = MAX(d, abs(((x>>16)&0xff)-((y>>16)&0xff)));
+
```

Comment:
These operations are not endian-safe. It is better to pass in a const uint8_t* and use array indexing to access the color components.